### PR TITLE
Interleaving deprecation

### PR DESCRIPTION
**Pull request by ericharmeling**
Fixes https://github.com/cockroachdb/docs/issues/9974.
Fixes https://github.com/cockroachdb/docs/issues/9622.
Fixes https://github.com/cockroachdb/docs/issues/9267.

This PR: 

- Improves the documentation on deprecation/removal of interleaved tabes/indexes in the 20.2 and 21.1 docs.
- Removes most of the references/recommendations for interleaved tables/indexes from the 20.2 and 21.1 docs. The exceptions being the actual `INTERLEAVE IN PARENT` syntax page and the 20.2.0 GA deprecation notices.

Note that this PR does not include guidance on querying `crdb_internal.interleaved*` tables (as suggested in https://github.com/cockroachdb/docs/issues/9974), as those tables (https://github.com/cockroachdb/cockroach/pull/61629 and https://github.com/cockroachdb/cockroach/pull/62076) have yet to make it to a release. When those PRs make it to a release branch and are part of a published release, we can update the crdb_internal page (also yet to be published: https://github.com/cockroachdb/docs/pull/8909), and then add a short example to `interleave-in-parent.html`.

### CODE REVIEW

**Review by vy-ton — COMMENTED** [review]
> Reviewed 52 of 52 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/docs/10154)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ericharmeling)

___
*[v20.2/backup.md, line 85 at r1](https://reviewable.io/reviews/cockroachdb/docs/10154#-MWeRjld0zJmh8EUaB6X:-MWeRjld0zJmh8EUaB6Y:bar829x) ([raw file](https://github.com/cockroachdb/docs/blob/65afce3c12e51f59a0ae08d6924619c35090bcfb/v20.2/backup.md#L85)):*
> ```Markdown
> Table with a [sequence](create-sequence.html) | The sequence it uses; however, this dependency can be [removed during the restore](restore.html#skip_missing_sequences).
> [Views](views.html) | The tables used in the view's `SELECT` statement.
> [Interleaved tables](interleave-in-parent.html) | The parent table in the [interleaved hierarchy](interleave-in-parent.html#interleaved-hierarchy).
> ```

We shouldn't delete this since backups will still need to account for dependencies on interleaved tables until the feature is fully removed
___
*[v20.2/cockroach-workload.md, line 177 at r1](https://reviewable.io/reviews/cockroachdb/docs/10154#-MWeScj_9KzIGe2FU3ug:-MWeScj_9KzIGe2FU3uh:bd9gk7f) ([raw file](https://github.com/cockroachdb/docs/blob/65afce3c12e51f59a0ae08d6924619c35090bcfb/v20.2/cockroach-workload.md#L177)):*
> ```Markdown
> `--histograms` | The file to write per-op incremental and cumulative histogram data to.<br><br>**Applicable command:** `run`
> `--init` | **Deprecated.** Use the `init` command instead.<br><br>**Applicable command:** `run`
> `--interleaved` | Use [interleaved tables](interleave-in-parent.html).<br><br>**Applicable commands:** `init` or `run`
> ```

https://github.com/cockroachdb/cockroach/issues/62609 tracks removing the flag. It does still exist today.
___
*[v20.2/cost-based-optimizer.md, line 169 at r1](https://reviewable.io/reviews/cockroachdb/docs/10154#-MWeT2Lb1x002PkI9Hyf:-MWeT2Lb1x002PkI9Hyg:b-hpnn8n) ([raw file](https://github.com/cockroachdb/docs/blob/65afce3c12e51f59a0ae08d6924619c35090bcfb/v20.2/cost-based-optimizer.md#L169)):*
> ```Markdown
> 
> {{site.data.alerts.callout_info}}
> With queries on [interleaved tables](interleave-in-parent.html), the optimizer might choose to use a merge join to perform a [foreign key](foreign-key.html) check when a lookup join would be more optimal.
> ```

Planning of the interleaved join was disabled by default in 20.2. Not sure why this doc compares merge. join vs lookup join for interleaved tables. 

Maybe @jordanlewis knows what we should highlight here if anything
___
*[v20.2/create-table.md, line 134 at r1](https://reviewable.io/reviews/cockroachdb/docs/10154#-MWeT_xI6wZ7xsJK-3CV:-MWeT_xI6wZ7xsJK-3CW:bxj3wyv) ([raw file](https://github.com/cockroachdb/docs/blob/65afce3c12e51f59a0ae08d6924619c35090bcfb/v20.2/create-table.md#L134)):*
> ```Markdown
> 
> {{site.data.alerts.callout_info}}
> Column families, interleavings, partitioning, and foreign key constraints
> ```

Probably can't delete this yet, since it's still true for 20.2 (i.e interleaved is deprecated but not removed yet)
___
*[v20.2/create-table-as.md, line 93 at r1](https://reviewable.io/reviews/cockroachdb/docs/10154#-MWeTmiXCFkbuUtLabUu:-MWeTmiXCFkbuUtLabUv:b-kepjyh) ([raw file](https://github.com/cockroachdb/docs/blob/65afce3c12e51f59a0ae08d6924619c35090bcfb/v20.2/create-table-as.md#L93)):*
> ```Markdown
> ## Limitations
> 
> Tables created with `CREATE TABLE ... AS` are not [interleaved](interleave-in-parent.html) with other tables.
> ```

Still true, prob shouldn't delete yet
___
*[v20.2/hash-sharded-indexes.md, line 24 at r1](https://reviewable.io/reviews/cockroachdb/docs/10154#-MWeU1On3nhcReXTUHTO:-MWeU1On3nhcReXTUHTP:bh26w6t) ([raw file](https://github.com/cockroachdb/docs/blob/65afce3c12e51f59a0ae08d6924619c35090bcfb/v20.2/hash-sharded-indexes.md#L24)):*
> ```Markdown
> 
> {{site.data.alerts.callout_info}}
> Hash-sharded indexes cannot be [interleaved](interleave-in-parent.html).
> ```

still true, can't delete yet
___
*[v20.2/import-into.md, line 329 at r1](https://reviewable.io/reviews/cockroachdb/docs/10154#-MWeU60TBX2PGpW39_A6:-MWeU60TBX2PGpW39_A7:b-kw3mdx) ([raw file](https://github.com/cockroachdb/docs/blob/65afce3c12e51f59a0ae08d6924619c35090bcfb/v20.2/import-into.md#L329)):*
> ```Markdown
> - After importing into an existing table, [constraints](constraints.html) will be un-validated and need to be [re-validated](validate-constraint.html).
> - Imported rows must not conflict with existing rows in the table or any unique secondary indexes.
> - `IMPORT INTO` works for only a single existing table, and the table must not be [interleaved](interleave-in-parent.html).
> ```

Still true
___
*[v20.2/interleave-in-parent.md, line 76 at r1](https://reviewable.io/reviews/cockroachdb/docs/10154#-MWeUTegDG2Bj54D7IzO:-MWeUTegDG2Bj54D7IzP:b-8ay1y5) ([raw file](https://github.com/cockroachdb/docs/blob/5b4df4eecbd4abab763b1501726fe87632b19ee8/v20.2/interleave-in-parent.md#L76)):*
> ```Markdown
> - Scans over tables or indexes with interleaved, child objects (i.e., interleaved tables or indexes) are much slower than scans over tables and indexes with no child objects, as the scans must traverse the parent object and all of its child objects.
> - Database schema changes are slower for interleaved objects and their parents than they are for non-interleaved objects and objects with no interleaved children. For example, if you add or remove a column to a parent or child table, CockroachDB must rewrite the entire interleaved hierarchy for that table and its parents/children.
> - [Internal benchmarks](https://github.com/cockroachdb/cockroach/issues/53455) have shown the performance benefits of interleaving tables and indexes are limited to a small number of use cases that do not justify the cost in engineering resources for maintenance.
> ```

Lets remove `that do not justify the cost in engineering resources for maintenance`
___
*[v20.2/interleave-in-parent.md, line 82 at r1](https://reviewable.io/reviews/cockroachdb/docs/10154#-MWeUqKB1592Oqxq6Ygo:-MWeUqKB1592Oqxq6Ygp:b32v57) ([raw file](https://github.com/cockroachdb/docs/blob/5b4df4eecbd4abab763b1501726fe87632b19ee8/v20.2/interleave-in-parent.md#L82)):*
> ```Markdown
> After [upgrading to v20.2](upgrade-cockroach-version.html), we recommend that you do the following:
> 
> - [Convert any existing interleaved tables to non-interleaved tables](#convert-interleaved-tables).
> ```

+1
___
*[v20.2/interleave-in-parent.md, line 91 at r1](https://reviewable.io/reviews/cockroachdb/docs/10154#-MWeUz0jCRntCyFn5VpM:-MWeUz0jCRntCyFn5VpN:bzds9y) ([raw file](https://github.com/cockroachdb/docs/blob/5b4df4eecbd4abab763b1501726fe87632b19ee8/v20.2/interleave-in-parent.md#L91)):*
> ```Markdown
> ### Convert interleaved tables
> 
> To convert an interleaved table to a non-interleaved table, issue an [`ALTER PRIMARY KEY`](alter-primary-key.html) statement on the table, specifying the existing primary key column(s) for the table, and no `INTERLEAVE IN PARENT` clause. Note that an `ALTER PRIMARY KEY` statement can only convert a child table if that table is not a parent. If your cluster has child tables that are also parents, you must start from the bottom of the interleaving hierarchy and work your way up (i.e., start with child tables that are not parents).
> ```

Let's mention that these schema changes are also online
___
*[v20.2/restore.md, line 143 at r1](https://reviewable.io/reviews/cockroachdb/docs/10154#-MWeVFGqDkWs3mcsahqt:-MWeVFGqDkWs3mcsahqu:b-b897ak) ([raw file](https://github.com/cockroachdb/docs/blob/65afce3c12e51f59a0ae08d6924619c35090bcfb/v20.2/restore.md#L143)):*
> ```Markdown
> Table with a [sequence](create-sequence.html) | The sequence.
> [Views](views.html) | The tables used in the view's `SELECT` statement.
> [Interleaved tables](interleave-in-parent.html) | The parent table in the [interleaved hierarchy](interleave-in-parent.html#interleaved-hierarchy).
> ```

Keep this
___
*[v20.2/sql-feature-support.md, line 171 at r1](https://reviewable.io/reviews/cockroachdb/docs/10154#-MWeVL8gDyqodF-nWlRZ:-MWeVL8gDyqodF-nWlR_:b-3ner1f) ([raw file](https://github.com/cockroachdb/docs/blob/65afce3c12e51f59a0ae08d6924619c35090bcfb/v20.2/sql-feature-support.md#L171)):*
> ```Markdown
> -----------|-----------|------|---------
>  Column families | ✓ | CockroachDB Extension | [Column Families documentation](column-families.html)
>  Interleaved tables | ✓ | CockroachDB Extension | [Interleaved Tables documentation](interleave-in-parent.html)
> ```

I would indicated Deprecated somehow. We still support existing interleaved tables until they are removed in  21.2

**Comment by vy-ton** [comment]
New migration guide looks great! We just need to be careful because interleaved tables are not yet removed, so all the docs related to their current behavior likely need to stick around for one more release.

We should be deleting/modifying anything that promotes their use though.

File: v20.2/interleave-in-parent.md [hunk 1]
```diff
@@ -111,41 +67,30 @@ For an example showing how to create tables that meet these criteria, see [Inter
...
-
-- Each child table's Primary Key must contain its parent table's Primary Key as a prefix (known as the **interleave prefix**).
-
-    For example, if the parent table's primary key is `(a INT, b STRING)`, the child table's primary key could be `(a INT, b STRING, c DECIMAL)`.
-
-    {{site.data.alerts.callout_info}}This requirement is enforced only by ensuring that the columns use the same data types. However, we recommend ensuring the columns refer to the same values by using the  <a href="foreign-key.html">Foreign Key constraint</a>.{{site.data.alerts.end}}
+## Deprecation
 
-- Interleaved tables cannot be the child of more than 1 parent table. However, each parent table can have many children tables. Children tables can also be parents of interleaved tables.
+Interleaving tables and indexes is deprecated in CockroachDB v20.2, and will be permanently disabled in a future release, for the following reasons:
```
Comment (vy-ton) [inline v20.2/interleave-in-parent.md:72]: @jordanlewis or @ajwerner might be interested in spot checking this

**Review by ericharmeling — COMMENTED** [review]
TFTR @vy-ton !

I think I've addressed all of your comments.

> We just need to be careful because interleaved tables are not yet removed, so all the docs related to their current behavior likely need to stick around for one more release.

Good point. I've reinstated some of the syntax/behavior notes that I removed. I will remove these again in 21.2, when interleaving is fully removed.

> We should be deleting/modifying anything that promotes their use though.

Agreed - all notes that fall into the category remain deleted/modified.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/docs/10154#-:-MWydOc6ArqLq8VSJeZc:b-d15bxj)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @jordanlewis, and @vy-ton)

___
*[v20.2/backup.md, line 85 at r1](https://reviewable.io/reviews/cockroachdb/docs/10154#-MWeRjld0zJmh8EUaB6X:-MWylWzL8GlXTzHvF_31:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/65afce3c12e51f59a0ae08d6924619c35090bcfb/v20.2/backup.md#L85)):*
<details><summary><i>Previously, vy-ton (Vy Ton) wrote…</i></summary><blockquote>

We shouldn't delete this since backups will still need to account for dependencies on interleaved tables until the feature is fully removed
</blockquote></details>

Done.
___
*[v20.2/cockroach-workload.md, line 177 at r1](https://reviewable.io/reviews/cockroachdb/docs/10154#-MWeScj_9KzIGe2FU3ug:-MWylZ931PuPhku-k76X:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/65afce3c12e51f59a0ae08d6924619c35090bcfb/v20.2/cockroach-workload.md#L177)):*
<details><summary><i>Previously, vy-ton (Vy Ton) wrote…</i></summary><blockquote>

https://github.com/cockroachdb/cockroach/issues/62609 tracks removing the flag. It does still exist today.
</blockquote></details>

Done.
___
*[v20.2/cost-based-optimizer.md, line 169 at r1](https://reviewable.io/reviews/cockroachdb/docs/10154#-MWeT2Lb1x002PkI9Hyf:-MWyd0d_AZ65dOsgL3nO:bs0sidb) ([raw file](https://github.com/cockroachdb/docs/blob/65afce3c12e51f59a0ae08d6924619c35090bcfb/v20.2/cost-based-optimizer.md#L169)):*
<details><summary><i>Previously, vy-ton (Vy Ton) wrote…</i></summary><blockquote>

Planning of the interleaved join was disabled by default in 20.2. Not sure why this doc compares merge. join vs lookup join for interleaved tables. 

Maybe @jordanlewis knows what we should highlight here if anything
</blockquote></details>

Yeah that doesn't make much sense... it probably should have been removed earlier.
___
*[v20.2/create-table.md, line 134 at r1](https://reviewable.io/reviews/cockroachdb/docs/10154#-MWeT_xI6wZ7xsJK-3CV:-MWyoDzD-lhhTDzj1KTl:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/65afce3c12e51f59a0ae08d6924619c35090bcfb/v20.2/create-table.md#L134)):*
<details><summary><i>Previously, vy-ton (Vy Ton) wrote…</i></summary><blockquote>

Probably can't delete this yet, since it's still true for 20.2 (i.e interleaved is deprecated but not removed yet)
</blockquote></details>

Done.
___
*[v20.2/create-table-as.md, line 93 at r1](https://reviewable.io/reviews/cockroachdb/docs/10154#-MWeTmiXCFkbuUtLabUu:-MWymopLERIHl_-RCn0g:b-cfikoz) ([raw file](https://github.com/cockroachdb/docs/blob/65afce3c12e51f59a0ae08d6924619c35090bcfb/v20.2/create-table-as.md#L93)):*
> Tables created with `CREATE TABLE ... AS` are not [interleaved](interleave-in-parent.html) with other tables.

I agree that this is still true, but I don't find it particularly relevant to anything. Especially if we don't want to encourage users to interleave tables.  `INTERLEAVE IN PARENT` has never been supported in `CREATE TABLE AS` AFAIK. 
___
*[v20.2/hash-sharded-indexes.md, line 24 at r1](https://reviewable.io/reviews/cockroachdb/docs/10154#-MWeU1On3nhcReXTUHTO:-MWynwSTEtak5UuVuMsQ:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/65afce3c12e51f59a0ae08d6924619c35090bcfb/v20.2/hash-sharded-indexes.md#L24)):*
<details><summary><i>Previously, vy-ton (Vy Ton) wrote…</i></summary><blockquote>

still true, can't delete yet
</blockquote></details>

Done.
___
*[v20.2/import-into.md, line 329 at r1](https://reviewable.io/reviews/cockroachdb/docs/10154#-MWeU60TBX2PGpW39_A6:-MWynx3N7NWooZnNGqqP:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/65afce3c12e51f59a0ae08d6924619c35090bcfb/v20.2/import-into.md#L329)):*
<details><summary><i>Previously, vy-ton (Vy Ton) wrote…</i></summary><blockquote>

Still true
</blockquote></details>

Done.
___
*[v20.2/interleave-in-parent.md, line 76 at r1](https://reviewable.io/reviews/cockroachdb/docs/10154#-MWeUTegDG2Bj54D7IzO:-MWydGWU4zCTyypgYy9r:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/5b4df4eecbd4abab763b1501726fe87632b19ee8/v20.2/interleave-in-parent.md#L76)):*
<details><summary><i>Previously, vy-ton (Vy Ton) wrote…</i></summary><blockquote>

Lets remove `that do not justify the cost in engineering resources for maintenance`
</blockquote></details>

Done.
___
*[v20.2/interleave-in-parent.md, line 91 at r1](https://reviewable.io/reviews/cockroachdb/docs/10154#-MWeUz0jCRntCyFn5VpM:-MWyy6vp8J1KpLHhOMqG:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/5b4df4eecbd4abab763b1501726fe87632b19ee8/v20.2/interleave-in-parent.md#L91)):*
<details><summary><i>Previously, vy-ton (Vy Ton) wrote…</i></summary><blockquote>

Let's mention that these schema changes are also online
</blockquote></details>

Done.
___
*[v20.2/restore.md, line 143 at r1](https://reviewable.io/reviews/cockroachdb/docs/10154#-MWeVFGqDkWs3mcsahqt:-MWyo-p56HoTE1SRc5aD:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/65afce3c12e51f59a0ae08d6924619c35090bcfb/v20.2/restore.md#L143)):*
<details><summary><i>Previously, vy-ton (Vy Ton) wrote…</i></summary><blockquote>

Keep this
</blockquote></details>

Done.
___
*[v20.2/sql-feature-support.md, line 171 at r1](https://reviewable.io/reviews/cockroachdb/docs/10154#-MWeVL8gDyqodF-nWlRZ:-MWyo0aS-fOsunh4KMgn:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/65afce3c12e51f59a0ae08d6924619c35090bcfb/v20.2/sql-feature-support.md#L171)):*
<details><summary><i>Previously, vy-ton (Vy Ton) wrote…</i></summary><blockquote>

I would indicated Deprecated somehow. We still support existing interleaved tables until they are removed in  21.2
</blockquote></details>

Done.

**Review by jordanlewis — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/docs/10154)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @vy-ton)

___
*[v20.2/cost-based-optimizer.md, line 169 at r1](https://reviewable.io/reviews/cockroachdb/docs/10154#-MWeT2Lb1x002PkI9Hyf:-MX8YlleDtQQwInaFpSr:b-ey4kd) ([raw file](https://github.com/cockroachdb/docs/blob/65afce3c12e51f59a0ae08d6924619c35090bcfb/v20.2/cost-based-optimizer.md#L169)):*
<details><summary><i>Previously, ericharmeling (Eric Harmeling) wrote…</i></summary><blockquote>

Yeah that doesn't make much sense... it probably should have been removed earlier.
</blockquote></details>

I think we can just remove this, I agree
___
*[v20.2/interleave-in-parent.md, line 72 at r1](https://reviewable.io/reviews/cockroachdb/docs/10154#-MWVgP1UNGGW2rbelvE4-r1-72:-MX8Z4FP8rlKe_CdRPkx:bb6ywmx) ([raw file](https://github.com/cockroachdb/docs/blob/5b4df4eecbd4abab763b1501726fe87632b19ee8/v20.2/interleave-in-parent.md#L72)):*
<details><summary><i>Previously, vy-ton (Vy Ton) wrote…</i></summary><blockquote>

@jordanlewis or @ajwerner might be interested in spot checking this
</blockquote></details>

LGTM

**Review by ianjevans — APPROVED** [review]

**Pull request merged**